pages/_document.js

```diff
@@ -1,27 +1,54 @@
 import Document, { Html, Head, Main, NextScript } from 'next/document'
+import { generateNonce } from '../lib/Utils.js'
 import Script from 'next/script'
 
 class MyDocument extends Document {
   static async getInitialProps(ctx) {
+    const nonce = generateNonce()
     const initialProps = await Document.getInitialProps(ctx)
-    return { ...initialProps }
+    //set csp
+    let csp = ``
+    csp += `base-uri 'self';`
+    csp += `form-action 'self';`
+    csp += `default-src 'self';`
+    csp += `script-src 'self' 'nonce-${nonce}' 'unsafe-eval';`
+    csp += `style-src 'self' https://fonts.googleapis.com 'unsafe-inline';`
+    csp += `img-src 'self' data: blob:;`
+    csp += `font-src 'self' https://fonts.gstatic.com data:;`
+    csp += `frame-ancestors 'self';`
+
+    if (ctx.res) {
+      ctx.res.setHeader('Content-Security-Policy', csp)
+    }
+
+    return { ...initialProps, nonce }
   }
 
   render() {
+    const { nonce } = this.props
     return (
       <Html className="no-js" dir="ltr" xmlns="http://www.w3.org/1999/xhtml">
-        <Head prefix="og:http://ogp.me/ns#"></Head>
+        <Head nonce={nonce} prefix="og:http://ogp.me/ns#"></Head>
         <body>
           <Main />
-          <NextScript />
+          <NextScript nonce={nonce} />
           <script
             async
             src="https://ajax.googleapis.com/ajax/libs/jquery/2.1.1/jquery.min.js"
+            nonce={nonce}
+            dangerouslySetInnerHTML={{
+              __html: `window.__webpack_nonce__ = "${nonce}"`,
+            }}
           ></script>
           {process.env.NEXT_PUBLIC_ADOBE_ANALYTICS_URL ? (
             <Script
+              id="1"
               strategy="beforeInteractive"
               src={process.env.NEXT_PUBLIC_ADOBE_ANALYTICS_URL}
+              nonce={nonce}
+              dangerouslySetInnerHTML={{
```

Comment:
It would be good if we can prevent using dangerouslySetInnerHTML.

One solution would be to purify the element with [DOMPurify](https://github.com/cure53/DOMPurify)

Found it on StackOverflow [here](https://stackoverflow.com/questions/29044518/safe-alternative-to-dangerouslysetinnerhtml)

If the team does not mind using `dangerouslySetInnerHTML`, I am ok with it.

Reply:
There are currently conversations about removing Adobe Analytics in favour of open telemetry. Might be worth removing this script all together

Reply:
Seems like the HTML didn't have much of a purpose so it's removed.